### PR TITLE
feat(consensus): dynamic validator sets

### LINKED ISSUE #310: Investigate merging tempo and reth datadirs
Tempo right now can store consensus and execution data in two different places, depending on how either config/command line arguments are configured. A node can be started in such a way that both end up under the same parent directory, but maybe it makes sense to make this a stronger requirement.

`xtask generate-config` shows how this is done manually by setting `--datadir` to write to the same place as consensus `Config.storage_directory`:

https://github.com/tempoxyz/tempo/blob/9bfb25275c02d1bab3a381e92fd349acbbdc627e/xtask/src/main.rs#L190-L195

### LINKED ISSUE #563: Store epoch length, initial BLS key, initial validator set public keys in chainspec
Right now, epoch lengths, the initial BLS key, and the initial validator set are passed to the node via config: https://github.com/tempoxyz/tempo/tree/6a303fbbee8df3208012e8a917deca79d04a1ff0/crates/commonware-node-config

This should be moved onchain, by writing it to the chainspec.

### LINKED ISSUE #883: consensus: write full peer set for an epoch to chain
### Describe the feature

As mentioned in https://github.com/tempoxyz/tempo/pull/585#discussion_r2516311152

> As discussed, we think there is a straightforward way to help new peers sync with a finalization cert on some boundary and either a proof of the contract state at that point (who will be the current peers) or by putting the info in the extra data field.

### Additional context

To implement this feature, the `PublicOutcome` type defined here needs to be updated with validator ingoing/outgoing addresses/hostnames (i.e. the values read from the `ValidatorConfig` smart contract, `incomingAddress` and `outgoingAddress`):
https://github.com/tempoxyz/tempo/blob/412186eed881a3e8d99256cd4c90dd3ee733ad83/crates/dkg-onchain-artifacts/src/lib.rs#L102-L106

Right now, the public outcome only contains the identifiers of the participants as determined by the DKG process (in the form of ed25519 public keys to verify their signatures).

We should probably just replace `participants` with this object here: https://github.com/tempoxyz/tempo/blob/412186eed881a3e8d99256cd4c90dd3ee733ad83/crates/commonware-node/src/dkg/manager/validators.rs#L113

### Sidequests:

Right now we use a bespoke encoding format. It probably makes sense to change this to RLP when writing onchain data.

### PR DESCRIPTION
**Pull request by SuperFluffy**
This patch allows dynamically adding and removing validators from Tempo by using the `ValidatorConfig` smart contract. The consensus layer is hardfork enabled, meaning that it only transitions to the new behavior after `allegrettoTime`.

There is a high level overview of the protocol provided in `docs/network-reconfiguration.md` provided with this patch.

### Configuration

1. consensus config was removed in favor of command line arguments, see `tempo node --help` and `crates/commonware-node/src/args.rs`
2 the public polynomial was moved to chainspec `"publicPolynomial"` entry
3. the initial set of peers was moved to chainspec `"validators"` entry
4. epoch length was moved to chainspec `"epochLength"` entry

### Adding and removing validators

1. where before `dealers == players` in each ceremony (and where both were defined through chainspec), the players in a new ceremony are now established by doing a `getValidators` call of the `ValidatorConfig` smart contract (technically, the validators returned by the contract call will be the players in the ceremony of the epoch after, giving them time to sync for one entire epoch; see the spec provided with this patch).
2. the p2p peers are similarly established through `getValidators`, where we track the peers of the last 3 epochs (current, parent, grandparent). The peers in epoch N is the union of the peers of `N`, `N+1` and `N+2`.
3. validators are removed by using the `changeValidatorStatus` smart contract call setting `active = false`. Similar to how adding validators work, this will remove the validator from first the `syncing players` list, and then from the `players` of the ceremony in the epoch after the next.

### DKG ceremonies and hardfork behavior

1. at startup, DKG ceremonies are run with the validator set specified in the chainspec. This matches the behavior before this patch (except that peers were provided through a config file)
2. when a new DKG ceremony is started (on the last height of the epoch) and if `boundary_block.timestamp >= allegretto_time`, the DKG manager reads the entries of the `ValidatorConfig` smart contract
4. if the entries match exactly the validators in the chainspec (i.e. `chainspec.validators == contract.validators` with both being a list `[(pubkey:socketaddr)]`), then a DKG ceremony according to post-hardfork logic is started.
5. from then on, validators can be added and removed by using the `ValidatorConfig` smart contract. 
6. in all other cases (timestamp is before allegretto time, the contract validators don't match), the pre-hardfork logic stays in place until the next epoch. This gives the operator time to fix the onchain contract state and provide the correct validator data.

Note 1: this requires an operator (admin account) to issue explicit `addValidator` calls to add all validators to the contract state.
Note 2: at the moment, a node always starts out in the pre-hardfork logic even if allegretto time is 0. This will be fixed in a followup.

### Xtasks

1. update xtask genesis generator to generate and inject validators, public polynomial into genesis/chainspec
2. rework the `generate-genesis` task by breaking out genesis-generation into its own set of arguments
3. rework the `generate-devnet` task in terms of this this new genesis-generation
4. remove `generate-config`

### e2e testing

1. rework e2e tests to get access to configs, execute smart contract calls from within the execution/tokio runtime
2. provide an e2e test for adding and removing a validator, a validator having lost its share and regaining it in a subsequent round, transition from pre- to post-allegretto logic

## Followup issues

https://github.com/tempoxyz/tempo/issues/562
https://github.com/tempoxyz/tempo/issues/565
https://github.com/tempoxyz/tempo/issues/792
https://github.com/tempoxyz/tempo/issues/884
https://github.com/tempoxyz/tempo/issues/924
https://github.com/tempoxyz/tempo/issues/928
https://github.com/tempoxyz/tempo/issues/930
https://github.com/tempoxyz/tempo/issues/933
https://github.com/tempoxyz/tempo/issues/1087

Closes #310 
Closes #563 
Closes #883

### CODE REVIEW

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+        ceremony::{self, Ceremony},
+        manager::{
+            actor::{DkgOutcome, pre_allegretto},
+            validators::ValidatorState,
+        },
+    },
+    epoch::{self, is_first_block_in_epoch},
+};
+
+const CURRENT_EPOCH_KEY: U64 = U64::new(0);
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:34]: these constants are also defined in pre_allegreto.rs, could be moved one level up

File: crates/commonware-node/src/dkg/manager/actor/mod.rs [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+    /// The actor configuration passed in when constructing the actor.
+    config: super::Config<TPeerManager>,
+
+    /// The runtime context passed in when constructing the actor.
+    context: ContextCell<TContext>,
+
+    /// The channel over which the actor will receive messages.
+    mailbox: mpsc::UnboundedReceiver<super::Message>,
+
+    /// Persisted information on the currently running ceremmony and its
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/actor/mod.rs]:
```suggestion
    /// Persisted information on the currently running ceremony and its
```

File: crates/commonware-node/src/dkg/manager/actor/mod.rs [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+
+    /// Persisted information on the current epoch for DKG ceremonies that were
+    /// started after the allegretto hardfork.
+    post_allegretto_metadatas: post_allegretto::Metadatas<ContextCell<TContext>>,
+
+    /// Persisted information on the current epoch for DKG ceremonies that were
+    /// started before the allegretto hardfork.
+    pre_allegretto_metadatas: pre_allegretto::Metadatas<ContextCell<TContext>>,
+
+    /// Information on the peers registered on the p2p peer mamnager for a given
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/actor/mod.rs]:
```suggestion
    /// Information on the peers registered on the p2p peer manager for a given
```

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+
+        // Remove it?
+        let dkg_outcome = self
+            .post_allegretto_metadatas
+            .dkg_outcome_metadata
+            .get(&DKG_OUTCOME_KEY.into())
+            .cloned()
+            .expect(
+                "when updating the current epoch state, there must be a DKG \
+                outcome of some ceremomny",
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs]:
```suggestion
                outcome of some ceremony",
```

File: crates/commonware-node/src/args.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
+//! Command line arguments for configuring the consensus layer of a tempo node.
+use std::{net::SocketAddr, path::PathBuf};
+
+/// Arguments for configuring the consensus layer of a tempo node.
```
Comment (fgimenez) [inline crates/commonware-node/src/args.rs]:
```suggestion
```

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+                    .expect("metadata must always be writable");
+            }
+        }
+
+        let mut ceremony = maybe_ceremony.take().expect(
+            "past this point a ceremony must always be defined; the only \
+                time a ceremony is not permitted to exist is exactly on the \
+                boundary; did the code after ensure that the ceremony is \
+                returned to its Option?",
+        );
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:146]: can this expect panic? in other words, does the code after ensure that the ceremony is returned to its Option?

File: crates/commonware-node/src/dkg/manager/validators.rs [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+    /// [`commonware_p2p::authenticated::lookup`]). If the inbound value is a
+    /// socket address, then the conversion is immediate. If is a domain name,
+    /// the domain name is resolved. If DNS resolution returns more than 1 value,
+    /// the last one is taken.
+    #[instrument(skip_all, fields(public_key = %self.public_key, inbound = %self.inbound), err)]
+    fn inbound_to_socket_addr(&self) -> eyre::Result<SocketAddr> {
+        let all_addrs = self
+            .inbound
+            .to_string()
+            .to_socket_addrs()
```
Comment (fgimenez) [inline crates/commonware-node/src/dkg/manager/validators.rs]: this is in the hot path when registering peers right? when using hostnames for peers this will involve dns lookups, could this be a problem?

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 1 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:34]: They share the same names and values but I want them to be confined (module-private) and not leak out.

File: crates/commonware-node/src/dkg/manager/validators.rs (hunk 7 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: It's in the hotpath, but finalization cannot conclude until the hostnames are resolved.

You are right though: this should be performed in a blocking task/extra thread.

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 6 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:146]: Yeah, see below.

It's very manual - I attempted to work around this with a guard, but the UX of that was awful.

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 1 above)
Comment (fgimenez) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:34]: sure, but they could be defined once in `crates/commonware-node/src/dkg/manager/actor/mod.rs` and still could be module private and be accessible pre and post allegretto?

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 6 above)
Comment (fgimenez) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:146]: great thx, could we update the expect message to make this clear?

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+        //
+        // This relies on the fact that the actor updates its tracked epoch
+        // state on the last block of the epoch.
+        if block_epoch != self.current_epoch_state().epoch() {
+            info!(
+                block_epoch,
+                actor_epoch = self.current_epoch_state().epoch(),
+                "block was for an epoch other than what the actor is currently tracking; ignoring",
+            );
+        }
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:106]: should we return early here ?

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:106]: Yes, good catch! I need to add a restart test for conditions post-allegretto && shutdown right after boundary-block finalization.

fad16919

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+                    participants: ceremony_outcome.participants,
+                    public,
+                    share,
+                },
+            )
+            .await
+            .expect("must always be able to persist the DKG outcome");
+
+        // Prune older ceremony.
+        if let Some(epoch) = current_epoch.checked_sub(1) {
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:217]: we used to prune -2, any special reason on why we prune the previous one now?

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+            epoch::RelativePosition::FirstHalf => {
+                let _ = ceremony.distribute_shares().await;
+                let _ = ceremony.process_messages().await;
+            }
+            epoch::RelativePosition::Middle => {
+                let _ = ceremony.process_messages().await;
+                let _ = ceremony.construct_intermediate_outcome().await;
+            }
+            epoch::RelativePosition::SecondHalf => {
+                let _ = ceremony.process_dealings_in_block(&block).await;
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs]: why don't we stop here if there's an error?

can we `inspect_err`  and log if we're not planning to exit here

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 9 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs:217]: Yeah, there is no point in keeping it around: we are at the end of of epoch `E`, having just finalized the ceremony that will give us the state for epoch `E+1`. There is nothing to do with the ceremony for `E-1`. I think technically we can even remove the ceremony for `E`.

The reason I shifted it here is to ensure that the ceremony for pre-allegretto is truly gone.

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs (hunk 10 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs]: We do provide an error event. The event is emitted in the scope of `process_dealings_in_block` where it belongs, see` #[instrument(err)]`

That's an unfortunate wart of how tracing works in Rust. :-( After the `process_dealings_in_block` ends, I have no way of adding an event to its scope.

File: crates/commonware-node/src/dkg/manager/validators.rs [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        // IMPORTANT: Starting with the syncing players to ensure that the
+        // latest address for a validator with a given pubkey is used.
+        // OrderedAssociated takes the first instance of a key it sees and
+        // drops the later instances.
+        self.syncing_players()
+            .iter_pairs()
+            .chain(self.players().iter_pairs())
+            .chain(self.dealers().iter_pairs())
+            .filter_map(|(pubkey, validator)| {
+                let addr = validator.inbound_to_socket_addr().ok()?;
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/validators.rs]: any DNS issue will result in us ignoring peers with domain names. unsure if this is truly desireable

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: Yes, that is a problem. I don't really know how to address this. Worse, the DNS problems can be localized to a single node.

Should we just spin on errors until all validators can be resolved?

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: To followup: note that the peers are only not registered on the peer manager (so we will not talk to them). We still track them and we still consider them potential players in a ceremony.

File: crates/commonware-node/src/args.rs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    #[arg(long = "consensus.signing-share")]
+    pub signing_share: Option<PathBuf>,
+
+    /// The TOML formatted file containing the bls12-381 threshold public key
+    /// polynomial and the list of peers.
+    #[arg(
+        long = "consensus.peers-and-public-polynomial",
+        required_unless_present_any = ["follow", "dev"],
+    )]
+    pub peers_and_public_polynomial: Option<PathBuf>,
```
Comment (klkvr) [inline crates/commonware-node/src/args.rs]: could this not be part of chainspec? also seems that we can keep that in genesis? just make the rule something like `last block of last non-Allegretto epoch must contain the dkg outcome`

that way for chains that were started before Allegretto we will just read it at transition

File: crates/commonware-node/src/dkg/ceremony/persisted.rs [hunk 13]
```diff
@@ -15,7 +15,7 @@ use super::IntermediateOutcome;
 /// Information on a ceremony that is persisted to disk.
 #[derive(Clone, Default, Debug, PartialEq, Eq)]
 pub(in crate::dkg) struct State {
-    pub(super) num_players: u64,
+    pub(super) num_players: u16,
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/ceremony/persisted.rs:18]: is this change backwards compatible?

Comment (joshieDo) [reply crates/commonware-node/src/dkg/ceremony/persisted.rs:18]: ok yes it is, its varint encoding

File: crates/commonware-node/src/dkg/manager/validators.rs [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+            inboundAddress,
+            outboundAddress,
+            ..
+        }: IValidatorConfig::Validator,
+    ) -> eyre::Result<Self> {
+        ensure!(
+            active,
+            "field `active` is set to false; this method should only be called \
+            for active validators"
+        );
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/validators.rs]: unsafe, decoding shouldnt do any checks imo

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: 8962ca13, good point. Removed also because it's redundant (filtering is done one level up).

File: crates/commonware-node/src/dkg/ceremony/persisted.rs (hunk 13 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/ceremony/persisted.rs:18]: yes :-)

File: crates/commonware-node/src/args.rs (hunk 12 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/args.rs]: All in chainspec as per [e307882](https://github.com/tempoxyz/tempo/pull/1039/commits/e307882c4ac131a54555654194a216702332535b)

I made it optional because writing some default values is pretty annoying.

File: crates/commonware-node/src/dkg/manager/validators.rs (hunk 7 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: This resolved itself: we no longer have hostnames in the inbound addresses as per [c9f055e](https://github.com/tempoxyz/tempo/pull/1039/commits/c9f055eca950f58186d0803cc1efd70a6f93911a) and  f9608776

File: crates/commonware-node/src/dkg/manager/validators.rs (hunk 11 above)
Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs]: Inbound addresses must be socket addresses as per [c9f055e](https://github.com/tempoxyz/tempo/pull/1039/commits/c9f055eca950f58186d0803cc1efd70a6f93911a) and  f9608776

File: crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs [hunk 15]
```diff
@@ -0,0 +1,609 @@
...
+                static validators = {static_validators:?}; \
+                on chain validators = {on_chain_validators:?}",
+            );
+        }
+
+        // NOTE: it would be good to compare the static validators addresses to
+        // what was read from the contract. But this would mean that some nodes
+        // transition to the dynamic validator regime while others don't,
+        // depending on how they are launched. Only the ed25519 keys are
+        // guaranteed to be the same across nodes.
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs]: is this whats happening above?

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/actor/post_allegretto.rs]: ty, added that check yesterdat and forgot the comment below it.  33bcf9c7

File: crates/dkg-onchain-artifacts/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,562 @@
...
+
+    fn read_cfg(
+        buf: &mut impl bytes::Buf,
+        _cfg: &Self::Cfg,
+    ) -> Result<Self, commonware_codec::Error> {
+        let n_players: u16 = UInt::read(buf)?.into();
+        let dealer = PublicKey::read(buf)?;
+        let dealer_signature = Signature::read(buf)?;
+        let epoch = UInt::read(buf)?.into();
+        let commitment = Public::<MinSig>::read_cfg(buf, &(quorum(n_players as u32) as usize))?;
```
Comment (joshieDo) [inline crates/dkg-onchain-artifacts/src/lib.rs:386]: If this is where IntermediateOutcome::decode ends up (i think?), and if a malicious validator builds this IntermediateOutcome with n_players == 0 , this will panic and disconnect every validator from the network

Comment (joshieDo) [reply crates/dkg-onchain-artifacts/src/lib.rs:386]: https://github.com/commonwarexyz/monorepo/blob/1ffaa1e6fb7c683b8b0ccc22adc8fe341499c24d/utils/src/lib.rs#L115-L118

Comment (joshieDo) [reply crates/dkg-onchain-artifacts/src/lib.rs:386]: need to validate n_players before we try the rest

File: crates/dkg-onchain-artifacts/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,562 @@
...
+        self.public.write(buf);
+    }
+}
+
+impl Read for PublicOutcome {
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl Buf, _cfg: &Self::Cfg) -> Result<Self, commonware_codec::Error> {
+        let epoch = UInt::read(buf)?.into();
+        let participants = Ordered::read_cfg(buf, &(RangeCfg::from(0..=usize::MAX), ()))?;
```
Comment (joshieDo) [inline crates/dkg-onchain-artifacts/src/lib.rs]: this needs to be capped, or else a malicious validator might craft an outcome that will try to allocate >130GB of memory

https://github.com/commonwarexyz/monorepo/blob/1ffaa1e6fb7c683b8b0ccc22adc8fe341499c24d/codec/src/types/vec.rs#L47-L48

Comment (joshieDo) [reply crates/dkg-onchain-artifacts/src/lib.rs]: we should review any other usage of `RangeCfg::from(0..=usize::MAX)` that would be passed by other validators

File: crates/dkg-onchain-artifacts/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,562 @@
...
+        let public =
+            Public::<MinSig>::read_cfg(buf, &(quorum(participants.len() as u32) as usize))?;
```
Comment (joshieDo) [inline crates/dkg-onchain-artifacts/src/lib.rs:124]: same issue as IntermediateOutcome, if participants is 0, this will panic every validator

File: crates/dkg-onchain-artifacts/src/lib.rs (hunk 17 above)
Comment (SuperFluffy) [reply crates/dkg-onchain-artifacts/src/lib.rs]: Capped this value at `u16::MAX`, still big, but manageable: ef1120b5

File: crates/dkg-onchain-artifacts/src/lib.rs (hunk 18 above)
Comment (SuperFluffy) [reply crates/dkg-onchain-artifacts/src/lib.rs:124]: Enforcing that `n_players != 0` now: ef1120b5

File: crates/dkg-onchain-artifacts/src/lib.rs (hunk 16 above)
Comment (SuperFluffy) [reply crates/dkg-onchain-artifacts/src/lib.rs:386]: Good catch: this is enforcing that `n_players != 0`: ef1120b5

File: crates/dkg-onchain-artifacts/src/lib.rs (hunk 18 above)
Comment (SuperFluffy) [reply crates/dkg-onchain-artifacts/src/lib.rs:124]: Finally fixed in [6610d25](https://github.com/tempoxyz/tempo/pull/1039/commits/6610d259d2908bcb689746f69f4358f9b1f865e3) - we now enforce that `RangeCfg::from(1..u16::MAX)` for the number of participants.

File: crates/commonware-node/src/dkg/manager/validators.rs [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+/// The inbound and outbound addresses stored herein are guaranteed to be of the
+/// form `<host>:<port>` for inbound, and `<ip>:<port>` for outbound. Here,
+/// `<host>` is either an IPv4 or IPV6 address, or a fully qualified domain name.
+/// `<ip>` is an IPv4 or IPv6 address.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct DecodedValidator {
+    pub(super) public_key: PublicKey,
+    pub(super) inbound: SocketAddr,
+    pub(super) outbound: SocketAddr,
+    pub(super) index: u64,
```
Comment (joshieDo) [inline crates/commonware-node/src/dkg/manager/validators.rs:286]: this index is not being used anywhere right? can we just remove it? Initially i was confused if this was somehow [wrongly] used in the ceremony.

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs:286]: It's not used, but I am displaying the validators in logs (see the `Display` impl) to show what exactly was read

Comment (joshieDo) [reply crates/commonware-node/src/dkg/manager/validators.rs:286]: okay probably cant remove it otherwise the new binary will have a bad time, can we just document this

Comment (joshieDo) [reply crates/commonware-node/src/dkg/manager/validators.rs:286]:
```suggestion
    /// Contract's internal array position; unused in node logic.
    pub(super) index: u64,
```

Comment (SuperFluffy) [reply crates/commonware-node/src/dkg/manager/validators.rs:286]: I exhaustively commented all the fields. :-)
